tools/i18n/update.js

```diff
@@ -0,0 +1,262 @@
+
+    processAttributes(element, strings);
+
+    var context = {
+        currentText: '',
+        ownerElement: element
+    };
+
+    element.children.forEach(function (child) {
+        processNode(child, strings, context);
+    });
+    processCurrentText(strings, context);
+}
+
+function processNode(node, strings, context) {
+    switch (node.type) {
+        case 'text':
+            context.currentText += node.data;
+            break;
+
+        case 'tag':
+            // If tag is not an inline tag, or it has attributes, finish off currentText and start clean
+            if (hasAttributes(node) || inlineTags.indexOf(node.name) == -1) {
+                processCurrentText(strings, context);
+                processAttributes(node, strings);
+                context.ownerElement = node;
+                node.children.forEach(function (child) {
+                    processNode(child, strings, context);
+                });
+                processCurrentText(strings, context);
+                return;
+            }
+            context.currentText += serializeNode(node);
+            break;
+
+        default:
+    }
+}
+
+function serializeNode(node) {
+    var docFragment = htmlparser2TreeAdapter.createDocumentFragment();
+    htmlparser2TreeAdapter.appendChild(docFragment, node);
+    return parse5.serialize(docFragment, {treeAdapter: htmlparser2TreeAdapter});
+}
+
+function processCurrentText(strings, context) {
+    var txt = context.currentText.trim();
+    context.currentText = '';
+
+    if (txt && shouldTranslate(txt)) {
+        var locId = context.ownerElement.attribs[LOC_ID_ATTRIB];
+        if (!locId) {
+            locId = applyId(context.ownerElement);
+        }
+        strings.push({type: 'text', id: locId, text: txt});
+    }
+}
+
+function shouldTranslate(text) {
+    // Remove stuff we wouldn't translate, and see if we have anything left
+
+    // Replace non-alpha
+    text = text.replace(nonAlphaRegex, ' ');
+
+    // Remove single characters
+    text = text.replace(singleCharRegex, ' ');
+
+    // Replace ignored words
+    ignoreWords.forEach(function (word) {
+        text = text.replace(word, ' ');
+    });
+
+    return !!text.trim();
+}
+
+function hasAttributes(element) {
+    return !!Object.getOwnPropertyNames(element.attribs).length;
+}
+
+function processAttributes(element, strings) {
+    if (!hasAttributes(element)) {
+        return null;
+    }
+
+    var attributes = element.attribs;
+    var locId = attributes[LOC_ID_ATTRIB];
+    translatedAttributes.forEach(function (translatedAttribute) {
+        var attValue = attributes[translatedAttribute];
+        if (attValue && shouldTranslate(attValue)) {
+            if (!locId) {
+                locId = applyId(element)
+            }
+            strings.push({type: 'attribute', id: translatedAttribute + '-' + locId, text: attValue});
+        }
+    });
+}
+
+function getId() {
+    return 'xxxxxxxx'.replace(/[xy]/g, function (c) {
```

Comment:
Oops, copy and paste error. I've replaced it with a less cool but faster version.